sched/Kconfig

```diff
@@ -398,6 +387,64 @@ config SMP_CALL
 
 endif # SMP
 
+config BMP
+	bool "Bound Multi-Processing (BMP)"
+	default n
+	depends on ARCH_HAVE_MULTICPU
+	depends on ARCH_HAVE_TESTSET
+	depends on ARCH_INTERRUPTSTACK != 0
+	select SPINLOCK
+	---help---
+	  Enables support for Bound Multi-Processing (BMP) on a multi-CPU
+	  platform. Bound multiprocessing provides the scheduling control
+	  of an asymmetric multiprocessing model, while preserving the
+	  hardware abstraction and management of symmetric multiprocessing.
+
+	  BMP is similar to SMP, but you can specify which processors a thread
+	  can run on. You can use both SMP and BMP on the same system, allowing
+	  some threads to migrate from one processor to another, while other
+	  threads are restricted to one or more processors.
+
+	  As with SMP, a single copy of the OS maintains an overall view of all
+	  system resources, allowing them to be dynamically allocated and shared
+	  among applications. But, during application initialization, a setting
+	  determined by the system designer forces all of an application's threads
+	  to execute only on a specified CPU.
+
+	  Compared to full, floating SMP operation, this approach offers several
+	  advantages:
+
+	  1. It eliminates the cache thrashing that can reduce performance in an SMP
+	  system by allowing applications that share the same data set to run
+	  exclusively on the same CPU.
+	  2. It offers simpler application debugging than SMP since all execution
+	  threads within an application run on a single CPU.
+	  3. It helps legacy applications that use poor techniques for synchronizing
+	  shared data to run correctly, again by letting them run on a single CPU.
+
+if BMP
+
+config BMP_NCPUS
```

Comment:
it's better to reuse SMP_NCPUS and SMP_CALL, but make BMP is a child option of SMP

Reply:
Thank you for the very-well-written and very helpful explanation!

Reply:
> it's better to reuse SMP_NCPUS and SMP_CALL, but make BMP is a child option of SMP

BMP_NCPUS is a little different from SMP_NCPUS. In AMP mode, SMP_NCPUS is always 1. Each CPU of BMP is actually an AMP core. If SMP_NCPUS is used to represent it in BMP, then all system resources will be dupped in multiple copies.

```
nuttx$ g CONFIG_SMP_NCPUS  sched/
sched/irq/irq_csection.c:58:volatile uint8_t g_cpu_nestcount[CONFIG_SMP_NCPUS];
sched/irq/irq.h:118:extern volatile uint8_t g_cpu_nestcount[CONFIG_SMP_NCPUS];
sched/sched/sched.h:73:#define is_idle_task(t)          ((t)->pid < CONFIG_SMP_NCPUS)
sched/sched/sched.h:192:extern dq_queue_t g_assignedtasks[CONFIG_SMP_NCPUS];
sched/sched/sched.h:199:extern FAR struct tcb_s *g_running_tasks[CONFIG_SMP_NCPUS];
sched/sched/sched_cpupause.c:83:  DEBUGASSERT(cpu != this_cpu() && (unsigned int)cpu < CONFIG_SMP_NCPUS);
sched/sched/sched_cpuselect.c:73:  for (i = 0; i < CONFIG_SMP_NCPUS; i++)
sched/sched/sched_processtimer.c:128:  for (i = 0; i < CONFIG_SMP_NCPUS; i++)
sched/sched/sched_sysinfo.c:103:  info->procs    = CONFIG_SMP_NCPUS;
sched/sched/sched_mergepending.c:220:       * Normally, this loop should execute no more than CONFIG_SMP_NCPUS

```